File: lintcode/merge-k-sorted-lists.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/merge-k-sorted-lists/

from linked_list import *
import heapq
import sys

class Solution:
    """
    @param lists: a list of ListNode
    @return: The head of one sorted list.
    """
    def mergeKLists(self, lists):
        # write your code here
        heap = []
        heapq.heapify(heap)
        for l in lists:
            if l is not None:
                heapq.heappush(heap, (l.val, l))

        if len(heap) == 0:
            return None
        if len(heap) == 1:
            val, node = heapq.heappop(heap)
            return node

        head = None
        p = None
        while len(heap) > 1:
            val, node = heapq.heappop(heap)
            nextnode = node.next
            node.next = None
            if not head:
                head = node
                p = head
            else:
                p.next = node
                p = node

            if nextnode is not None:
                heapq.heappush(heap, (nextnode.val, nextnode))

        if len(heap):
            val, node = heapq.heappop(heap)
            if not head:
                head = node
            else:
                p.next = node

        return head

if __name__ == '__main__':
    head1 = create_linked_list([2,4])
    head2 = create_linked_list([])
    head3 = create_linked_list([-1])
    lists = [head1, head2, head3]
    s = Solution()
    head = s.mergeKLists(lists)
    print_linked_list(head)

File: leetcode-2/merge-k-sorted-lists.cc
// Question: https://leetcode.com/problems/merge-k-sorted-lists/

#include "headers.h"

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

struct cmp {
    bool operator () (ListNode* n1, ListNode* n2) {
        return n1->val > n2->val;  // 大的放后面，小的先出队。
    }
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) {
            return NULL;
        }
        if (lists.size() == 1) {
            return lists[0];
        }
        priority_queue<ListNode*, vector<ListNode*>, cmp> pq;
        for (auto node : lists) {
            if (node) {
                pq.push(node);
            }
        }

        ListNode dummy(0);
        auto cursor = &dummy;
        while (!pq.empty()) {
            auto node = pq.top();
            pq.pop();
            if (node->next) {
                pq.push(node->next);
            }

            node->next = cursor->next;
            cursor->next = node;
            cursor = node;
        }
        return dummy.next;
    }
};

int main(int argc, char** argv) {
}
